pkg/operation/botanist/component/deploy.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package component
+
+import "context"
+
+// OpDestroy creates a DeployWaiter which calls Destroy instead of create
+// and WaitCleanup instead of Wait
+func OpDestroy(dw ...DeployWaiter) DeployWaiter {
+	return &deploy{
+		dw:     dw,
+		invert: true,
+		wait:   false,
+	}
+}
+
+// OpDestroyAndWait creates a Deployer which calls Destroy instead of create
+// and waits for destruction.
+func OpDestroyAndWait(dw ...DeployWaiter) Deployer {
+	return &deploy{
+		dw:     dw,
+		invert: true,
+		wait:   true,
+	}
+}
+
+// OpWaiter creates a Deployer which calls waits for each operation.
+func OpWaiter(dw ...DeployWaiter) Deployer {
+	return &deploy{
+		dw:     dw,
+		invert: false,
+		wait:   true,
+	}
+}
+
+// NoOp does nothing
+func NoOp() DeployWaiter { return &deploy{} }
+
+type deploy struct {
+	invert bool
+	wait   bool
+	dw     []DeployWaiter
+}
+
+func (d *deploy) Deploy(ctx context.Context) error {
+	if d.invert {
```

Comment:
What is the purpose of this? There is already a `Destroy` function - why would I set `invert=true` and call `Deploy` instead of calling `Destroy` directly?

If I look at above functions `OpDestroy`, etc. then I assume that the `Destroy` function of the `Deployer` interface is never called? I'm confused, tbh. Can we get rid of the `invert` field and make it rather explicit what is called?

Reply:
The principle is simple - if a component was installed in reconcilationA then removed in reconcilationB (remove DNS entries, DNSProviders and so), the reconciler has to explicitly know the previous state and call `Destroy` explicitly to cleanup left over resources. With this, you can construct a component based on the actual state you want to have in the system.

Example with some pseudo code - when constructing the botanist you already know the desired state of the `Foo` component - to deploy or to remove. With this the create will simply call `Foo.Deploy()` and this would clean up old resources.

```go

type Botanist struct {
	Foo component.DeployWaiter
}

func NewBotanist() *Botanist {
	b := &Botanist{}
	if fooEnabled {
		b.Foo = NewFoo()
	} else {
		b.Foo = OpDestroy(NewFoo())
	}
	return b
}


// Create step 
_ = g.Add(flow.Task{
		Name:         "Ensure Foo",
		Fn:           flow.TaskFn(botanist.Foo.Deploy)
})
```

Reply:
I see... but wouldn't the following be as convenient as above while allowing to remove the `invert` logic?

```golang
type Botanist struct {}

func NewBotanist() *Botanist {
	b := &Botanist{}

	var ensureFooTask flow.TaskFn
	if fooEnabled {
		ensureFooTask = flow.TaskFn(NewFoo().Deploy)
	} else {
		ensureFooTask = flow.TaskFn(NewFoo().Destroy)
	}

	return b
}


// Create step 
_ = g.Add(flow.Task{
		Name:         "Ensure Foo",
		Fn:           ensureFooTask,
})
```

Reply:
I also tried playing with this, but the problem with this approach is that in some cases a component have to be deployed, then removed (deploying kube-apiserver for example when deleting).

Reply:
Hm, can you elaborate more why this does not work, maybe again with pseudo-code (in comparison to how it work with the current approach)?

Reply:
I've renamed the field to `deleteOnly` to make more clear.